src/transformers/modeling_utils.py

```diff
@@ -1465,7 +1798,14 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             if torch_dtype is not None:
                 if isinstance(torch_dtype, str):
                     if torch_dtype == "auto":
-                        torch_dtype = next(iter(state_dict.values())).dtype
+                        if is_sharded and "dtype" in sharded_metadata:
+                            torch_dtype = sharded_metadata["dtype"]
+                        elif not is_sharded:
+                            torch_dtype = next(iter(state_dict.values())).dtype
+                        else:
+                            one_state_dict = load_state_dict(resolved_archive_file)
+                            torch_dtype = next(iter(one_state_dict.values())).dtype
+                            del one_state_dict  # free CPU memory
```

Comment:
Adapting the previous logic for sharded checkpoints where we don't have a state dict yet:
- the user may have put the proper `dtype` in the metadata of the index of the checkpoint
- otherwise load the first state dict and get the first dype, which is the same as what happens for non-sharded checkpoints.